compiler/optimizer/Structure.cpp

```diff
@@ -184,7 +186,788 @@ void TR_RegionStructure::cleanupAfterEdgeRemoval(TR::CFGNode *node)
+         return;
+
+      TR_RegionStructure * const parent = region->getParent();
+      if (parent == NULL)
+         return;
+
+      // Try to find a remaining successor of node in this region.
+      TR_SuccessorIterator sit(node);
+      for (auto e = sit.getFirst(); e != NULL; e = sit.getNext())
+         {
+         if (!region->isExitEdge(e))
+            return;
+         }
+
+      // None of this node's remaining successors are in this region, so it
+      // unconditionally exits, and we can move it into the parent region.
+      // However, we'd like to do so only if all of the remaining successors
+      // are outside the innermost containing loop. Otherwise we'll spend time
+      // dissolving acyclic regions, or worse, invalidate structure, for no
+      // useful purpose.
+      //
+      // This test is delayed until we know that all outgoing edges are exit
+      // edges in the hope that it will be unnecessary to build the set of node
+      // numbers contained in the loop.
+      //
+      // Note that if (region == loop), then we already know that all
+      // successors are outside loop, because they're outside region.
+      if (region != loop)
+         {
+         TR_BitVector &loopBlocks = regionContents(loop);
+         for (auto e = sit.getFirst(); e != NULL; e = sit.getNext())
+            {
+            if (loopBlocks.isSet(e->getTo()->getNumber()))
+               return;
+            }
+         }
+
+      // This node doesn't belong in the loop!
+      if (!performTransformation(_comp,
+            "%sMoving unconditional exit node %d "
+            "from region %d:%p into parent %d:%p\n",
+            OPT_DETAILS,
+            node->getNumber(),
+            region->getNumber(),
+            region,
+            parent->getNumber(),
+            parent))
+         return;
+
+      if (_comp->getOption(TR_ExtractExitsByInvalidatingStructure))
+         {
+         // There is a node that would be extracted; just kill structure instead
+         if (_trace)
+            traceMsg(_comp, "invalidating structure instead of fixing it\n");
+
+         // FIXME this should call invalidateStructure(), but doing so does not
+         // clear out blocks' getStructureOf(), and not everything that looks
+         // at getStructureOf() checks beforehand that structure is valid. So
+         // at the moment, invalidateStructure() can cause subsequent
+         // optimizations to dereference dangling pointers.
+         _cfg->setStructure(NULL);
+         return;
+         }
+
+      moveNodeIntoParent(node, region, parent);
+      removeContentsFromRegion(initialStructure, region);
+
+      if (_trace && _comp->getOutFile() != NULL)
+         {
+         traceMsg(_comp, "Structure after moving node into parent:\n<structure>\n");
+         _comp->getDebug()->print(_comp->getOutFile(), _cfg->getStructure(), 0);
+         traceMsg(_comp, "</structure>\n");
+         }
+
+#ifdef DEBUG
```

Comment:
This looks loop invariant unless I am reading the nesting wrong - perhaps it should move out?

Reply:
The value of `checkStructure` could be determined outside the loop, but I think it's clearer next to its use here, and the overhead of reading an option should be negligible in this context

Reply:
Given the distance from the loop header I can buy this - in general we do want to avoid having too many little things like this creep in since they do eventually add up. I do agree, however, in this situation the clarity probably warrants the current placement.